optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py

```diff
@@ -0,0 +1,331 @@
+            self.error_exit(args)
+
+        parameters = [(key, value.split(', '))
+                      for key, value in config.items('parameters')]
+        constraints = [(key, value.split(', '))
+                       for key, value in config.items('constraints')]
+        order = args.cit_order_of_combinations
+        cit = Cit(parameters, constraints, order)
+        self.headers, self.variants = cit.combine()
+
+    @staticmethod
+    def error_exit(args):
+        if args.subcommand == 'run':
+            sys.exit(exit_codes.AVOCADO_JOB_FAIL)
+        else:
+            sys.exit(exit_codes.AVOCADO_FAIL)
+
+    def __iter__(self):
+        if self.variants is None:
+            return
+
+        variant_ids = []
+        for variant in self.variants:
+            base_id = "-".join([variant.get(key) for key in self.headers])
+            variant_ids.append(base_id + '-' +
+                               hashlib.sha1(base_id).hexdigest()[:4])
+
+        for vid, variant in itertools.izip(variant_ids, self.variants):
+            variant_tree_nodes = []
+            for key, val in variant.items():
+                variant_tree_nodes.append(TreeNode(key, {key: val}))
+
+            yield {"variant_id": vid,
+                   "variant": variant_tree_nodes,
+                   "paths": self.parameter_path}
+
+    def __len__(self):
+        return sum(1 for _ in self)
+
+    def update_defaults(self, defaults):
+        pass
+
+    def to_str(self, summary, variants, **kwargs):
+        """
+        Return human readable representation
+
+        The summary/variants accepts verbosity where 0 means silent and
+        maximum is up to the plugin.
+
+        :param summary: How verbose summary to output (int)
+        :param variants: How verbose list of variants to output (int)
+        :param kwargs: Other free-form arguments
+        :rtype: str
+        """
+        if not self.variants:
+            return ""
+
+        out = []
+        verbose = variants > 1
+        out.append("CIT Variants (%i):" % len(self))
+        for variant in self:
+            out.append('%sVariant %s:    %s' % ('\n' if verbose else '',
+                                                variant["variant_id"],
+                                                self.parameter_path))
+            if not verbose:
+                continue
+            env = set()
+            for node in variant["variant"]:
+                for key, value in iteritems(node.environment):
+                    origin = node.environment.origin[key].path
+                    env.add(("%s:%s" % (origin, key), str(value)))
+            if not env:
+                return out
+            fmt = '    %%-%ds => %%s' % max([len(_[0]) for _ in env])
+            for record in sorted(env):
+                out.append(fmt % record)
+
+        return "\n".join(out)
+
+
+class Cit(object):
+
+    MATRIX_ROW_SIZE = 20
+    MAX_INTERACTIONS = 15
```

Comment:
Hmm, do we expect users to experiment with those values?

Reply:
At this point, I don't think so.